byterun/weak.c

```diff
@@ -155,240 +189,302 @@ static void do_set (value ar, mlsize_t offset, value v)
-    res = None_val;
+    return 0;
   }else{
     elt = Field (ar, offset);
     if (caml_gc_phase == Phase_mark && Must_be_Marked_during_mark(elt)){
       caml_darken (elt, NULL);
     }
-    res = caml_alloc_small (1, Some_tag);
-    Field (res, 0) = elt;
+    *key = elt;
+    Assert_not_dead_value(elt);
+    return 1;
   }
-  CAMLreturn (res);
 }
 
-CAMLprim value caml_weak_get (value ar, value n){
+CAMLprim value caml_ephe_get_key (value ar, value n)
+{
+  value data;
+  return optionalize(caml_ephemeron_get_key(ar,Long_val(n),&data),&data);
+}
+
+CAMLprim value caml_weak_get (value ar, value n)
+{
   return caml_ephe_get_key(ar, n);
 }
 
-CAMLprim value caml_ephe_get_data (value ar)
+CAMLexport int caml_ephemeron_get_data (value ar, value *data)
 {
-  CAMLparam1 (ar);
-  mlsize_t offset = 1;
-  CAMLlocal2 (res, elt);
-                                                   Assert (Is_in_heap (ar));
-  elt = Field (ar, offset);
+  value elt;
+  Assert_valid_ephemeron(ar);
+
   if(caml_gc_phase == Phase_clean) caml_ephe_clean(ar);
+  elt = Field (ar, CAML_EPHE_DATA_OFFSET);
   if (elt == caml_ephe_none){
-    res = None_val;
+    return 0;
   }else{
     if (caml_gc_phase == Phase_mark && Must_be_Marked_during_mark(elt)){
       caml_darken (elt, NULL);
     }
-    res = caml_alloc_small (1, Some_tag);
-    Field (res, 0) = elt;
+    *data = elt;
+    Assert_not_dead_value(elt);
+    return 1;
   }
-  CAMLreturn (res);
 }
 
-#undef Setup_for_gc
-#undef Restore_after_gc
+CAMLprim value caml_ephe_get_data (value ar)
+{
+  value data;
+  return optionalize(caml_ephemeron_get_data(ar,&data),&data);
+}
 
-CAMLprim value caml_ephe_get_key_copy (value ar, value n)
+
+static inline void copy_value(value src, value dst)
 {
-  CAMLparam2 (ar, n);
-  mlsize_t offset = Long_val (n) + 2;
-  CAMLlocal2 (res, elt);
-  value v;  /* Caution: this is NOT a local root. */
-                                                   Assert (Is_in_heap (ar));
-  if (offset < 1 || offset >= Wosize_val (ar)){
-    caml_invalid_argument ("Weak.get_copy");
+  if (Tag_val (src) < No_scan_tag){
+    mlsize_t i;
+    for (i = 0; i < Wosize_val (src); i++){
+      value f = Field (src, i);
+      if (caml_gc_phase == Phase_mark && Must_be_Marked_during_mark(f)){
+        caml_darken (f, NULL);
+      }
+      caml_modify (&Field (dst, i), f);
+    }
+  }else{
+    memmove (Bp_val (dst), Bp_val (src), Bosize_val (src));
   }
+};
 
-  if (is_ephe_key_none(ar, offset)) CAMLreturn (None_val);
-  v = Field (ar, offset);
-  if (Is_block (v) && Is_in_heap_or_young(v)) {
-    elt = caml_alloc (Wosize_val (v), Tag_val (v));
-          /* The GC may erase or move v during this call to caml_alloc. */
+CAMLexport int caml_ephemeron_get_key_copy(value ar, mlsize_t offset,
+                                           value *key)
+{
+  CAMLparam1(ar);
+  value elt = Val_unit, v; /* Caution: they are NOT a local root. */
+  Assert_valid_offset(ar,offset);
+
+  offset += CAML_EPHE_FIRST_KEY;
+
```

Comment:
(At the moment, is this relying on some property like: values may only be truncated, so the recursion between finalisers added on the key and this function is always well-founded? (!) )